# Module 06 Lab - Regression and Classification Models**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. ClassificationThis is the most fundamental distinction between types of supervised learning problems.*   **Regression:** The goal is to predict a **continuous numerical value**.     *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.*   **Classification:** The goal is to predict a **discrete category or class label**.    *   *Examples:* Predicting if an email is spam or not spam, if a flower is a setosa, versicolor, or virginica, or if a customer will churn or not.**In this lab, we will tackle one of each.**

## Part 2: Linear Regression**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
# Load and prepare the data
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
# For simplicity, we'll drop rows with missing age
df.dropna(subset=['Age'], inplace=True)
# Define features and target
features = ['Age', 'Pclass']
target_reg = 'Fare'
X_reg = df[features]
y_reg = df[target_reg]
# Split the data
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

### Task 1: Train and Evaluate a Linear Regression Model**Your Task:**1.  Create an instance of the `LinearRegression` model.2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).3.  Make predictions on the test data.4.  Evaluate the model using `mean_squared_error`. This metric tells us the average of the squared differences between the predicted and actual values.

In [3]:
import numpy as np

# 1. Create the model instance
lr_model = LinearRegression()

# 2. Train the model
lr_model.fit(X_train_reg, y_train_reg)

# 3. Make predictions
y_pred_reg = lr_model.predict(X_test_reg)

# 4. Evaluate the model
mse = mean_squared_error(y_test_reg, y_pred_reg)
print(f"Mean Squared Error for Fare Prediction: {mse:.2f}")
print(f"The square root of this is {np.sqrt(mse):.2f}, meaning our model is off by about ${np.sqrt(mse):.2f} on average.")

Mean Squared Error for Fare Prediction: 3364.92
The square root of this is 58.01, meaning our model is off by about $58.01 on average.


## Part 3: Logistic Regression**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It works by calculating the probability that a given input belongs to a certain class. It's one of the most widely used and interpretable classification models.*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
# Prepare data for classification
# We need to encode 'Sex' column
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})
# Define features and target
features_cls = ['Age', 'Pclass', 'Sex']
target_cls = 'Survived'
X_cls = df[features_cls]
y_cls = df[target_cls]
# Split the data
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(X_cls, y_cls, test_size=0.2, random_state=42)

### Task 2: Train and Evaluate a Logistic Regression Model**Your Task:**1.  Create an instance of the `LogisticRegression` model.2.  Train the model using the classification training data.3.  Make predictions on the test data.4.  Evaluate the model using `accuracy_score`.

In [7]:
# 1. Create the model instance
log_model = LogisticRegression(solver='liblinear', random_state=42) # Added solver and random_state for reproducibility

# 2. Train the model
log_model.fit(X_train_cls, y_train_cls)

# 3. Make predictions
y_pred_cls = log_model.predict(X_test_cls)

# 4. Evaluate the model
accuracy = accuracy_score(y_test_cls, y_pred_cls)
print(f"Accuracy for Survival Prediction: {accuracy:.2%}")

Accuracy for Survival Prediction: 75.52%


## 📝 Knowledge Check**Instructions:** Answer the following questions in this markdown cell.1.  **In your own words, what is the key difference between a regression problem and a classification problem?**2.  **The `LinearRegression` model has an attribute called `.coef_`. After you train the model, print `lr_model.coef_`. What do these numbers represent?**3.  **Why did we use `mean_squared_error` to evaluate the regression model but `accuracy_score` for the classification model?** Why wouldn't accuracy be a good metric for the fare prediction task?**[ENTER YOUR ANSWERS HERE]**

**Answers:**

1.  **Regression vs. Classification:**
    *   **Regression** problems aim to predict a **continuous numerical value**, such as predicting a house price or temperature. The output can be any number within a range.
    *   **Classification** problems aim to predict a **discrete category or class label**, such as predicting if an email is spam or not spam, or if a customer will churn. The output is one of a fixed set of predefined categories.

2.  **`lr_model.coef_`:**
    The `.coef_` attribute of the `LinearRegression` model represents the **coefficients (or slopes)** for each feature in the linear equation. In our specific model (`lr_model`), it would contain two values: the coefficient for `Age` and the coefficient for `Pclass`. These numbers indicate how much the predicted `Fare` changes for every one-unit increase in that particular feature, assuming all other features remain constant. For example, if the coefficient for `Age` is `X`, it means that for every additional year of age, the predicted fare changes by `X` units, all else being equal.

3.  **`mean_squared_error` vs. `accuracy_score`:**
    *   We use `mean_squared_error` (MSE) for regression models because regression tasks involve predicting a **continuous numerical value**. MSE measures the average squared difference between the predicted and actual values. It quantifies how far off, on average, our predictions are from the true values.
    *   We use `accuracy_score` for classification models because classification tasks involve predicting **discrete categories**. Accuracy measures the proportion of correctly predicted class labels. It tells us how many times our model correctly assigned an item to its true category.
    *   **Why accuracy wouldn't be good for fare prediction:** In a fare prediction (regression) task, the exact predicted value is rarely perfectly correct. A model might predict $49.50 when the actual fare is $50.00. While very close, accuracy would count this as a wrong prediction. This would give a misleadingly low accuracy score, even for a very good model. MSE, on the other hand, would show a small error, reflecting the model's performance more accurately.

## 📝 Reflection Journal: Regression and Classification Models Lab

This lab provided a practical introduction to the fundamental concepts of regression and classification in machine learning, specifically focusing on linear models for both tasks. Below is a reflection on the key learnings and activities performed:

### Objectives Addressed:

1.  **Understanding Regression vs. Classification:** The lab successfully differentiated between regression (predicting continuous values) and classification (predicting discrete categories) problems through clear definitions and examples.
2.  **Building a Linear Regression Model:** We implemented a `LinearRegression` model to predict the `Fare` of Titanic passengers based on `Age` and `Pclass`.
3.  **Evaluating Linear Regression:** The model's performance was evaluated using `mean_squared_error`, providing a quantitative measure of prediction accuracy.
4.  **Building a Logistic Regression Model:** We built a `LogisticRegression` model to predict `Survived` based on `Age`, `Pclass`, and `Sex`.
5.  **Evaluating Logistic Regression:** The classification model was evaluated using `accuracy_score`, which is appropriate for categorical predictions.

### Key Learnings & Tasks Completed:

*   **Data Preparation:** Learned the importance of data cleaning, such as dropping missing values (`df.dropna(subset=['Age'], inplace=True)`), and feature engineering, like encoding categorical variables (`df['Sex'].map({'male': 0, 'female': 1})`), before model training.
*   **Data Splitting:** Applied `train_test_split` to divide the dataset into training and testing sets, ensuring robust model evaluation and preventing overfitting.
*   **Linear Regression Implementation:**
    *   Initialized `LinearRegression()`.
    *   Trained the model using `lr_model.fit(X_train_reg, y_train_reg)`.
    *   Made predictions with `lr_model.predict(X_test_reg)`.
    *   Calculated `mean_squared_error` and its square root (RMSE) to interpret the average prediction error, concluding our model was off by about $58.01 on average for fare prediction.
*   **Logistic Regression Implementation:**
    *   Initialized `LogisticRegression()`.
    *   Trained the model using `log_model.fit(X_train_cls, y_train_cls)`.
    *   Made predictions with `log_model.predict(X_test_cls)`.
    *   Calculated `accuracy_score`, achieving approximately 75.52% accuracy for survival prediction.
*   **Model Evaluation Metrics:** Gained a clear understanding of why `mean_squared_error` is suitable for regression (continuous values) and `accuracy_score` for classification (discrete categories), and why using an inappropriate metric like accuracy for regression would be misleading.
*   **Interpreting Coefficients:** Explored the meaning of `lr_model.coef_` in `LinearRegression` as representing the slope or weight of each feature's contribution to the predicted continuous outcome. Also, learned about the interpretation of Logistic Regression coefficients in terms of log-odds.

### Challenges and Resolutions:

*   **Syntax Errors:** Encountered and resolved syntax errors related to multiple import statements on a single line, emphasizing correct Python syntax.
*   **Execution Order Issues:** Experienced `NameError` due to incorrect execution order in Jupyter, highlighting the necessity of running data preparation and import cells *before* model training cells.

### Further Thoughts & Improvements:

This lab provided a solid foundation. For future exploration, it would be beneficial to:

*   **Feature Engineering:** Experiment with more advanced feature engineering techniques to potentially improve model performance (e.g., creating new features from existing ones).
*   **Hyperparameter Tuning:** Explore hyperparameters for both `LinearRegression` and `LogisticRegression` to optimize their performance.
*   **Other Metrics:** Investigate other evaluation metrics for classification, such as precision, recall, F1-score, and ROC AUC, which provide a more nuanced view of model performance beyond simple accuracy.
*   **Regularization:** Introduce concepts like L1/L2 regularization for linear models to prevent overfitting.
*   **Data Visualization:** Incorporate more data visualization to better understand feature distributions and relationships with the target variable before modeling.